### PR TITLE
Implement sce_em_set

### PR DESCRIPTION
**Pull request by IntelOrca**
Implements `sce_em_set` opcode, which sets up a new enemy spawn for a room. Some additional code is implemented:
* Initialization of game enemy routine table
* Initialization of room enemy routine table
* Zombie dog routine top level and dead state

### CODE REVIEW

File: src/enemy.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        EM_CLAIRE = 81,
+        EM_LEON_BANDAGED = 84,
+        EM_CLAIRE_NO_JACKET = 85,
+        EM_LEON_TANKTOP = 88,
+        EM_CLAIRE_COWGIRL = 89,
+        EM_LEON_LEATHER = 90,
+        EM_NONE = 0xFF,
+    };
+
+    bool is_enemy_dead(uint8_t globalId);
```
Comment (elmarsan) [inline src/enemy.h:79]: What do you think about using [[nodiscard]] here and in `em_kind_search` ?

File: src/interop.hpp [hunk 2]
```diff
@@ -400,6 +402,28 @@ namespace openre::interop
...
     bool operator!=(const save_state& lhs, const save_state& rhs);
 
+    class memory_comparer
+    {
+    private:
+        uintptr_t _begin = 0;
+        uintptr_t _end = 0;
+        save_state _pre;
+        save_state _first;
+        save_state _second;
```
Comment (elmarsan) [inline src/interop.hpp:412]: Should this `save_state` be a class member if it's only used in the `log `function for logging purposes?

File: src/openre.cpp [hunk 3]
```diff
@@ -328,6 +330,16 @@ namespace openre
         auto addr = gGameTable.flag_groups[static_cast<uint32_t>(group)];
         bitarray_set(addr, index, value);
     }
+
+    void* work_alloc(size_t len)
+    {
+        auto mem = gGameTable.mem_top;
+        gGameTable.mem_top = reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(mem) + len);
+#ifdef DEBUG
+        std::memset(mem, 0xCD, len);
```
Comment (elmarsan) [inline src/openre.cpp:339]: Can you explain me why are you doing this, please?

File: address_map.txt [hunk 4]
```diff
@@ -118,11 +124,54 @@ declare     tpage                   u8
...
+declare     var_188                 s32         0x188
+declare     var_18C                 s32         0x18C
+declare     pSa_dat                 void*       0x1BC
+declare     var_1C0                 u8          0x1C0
+declare     var_1C2                 s16         0x1C2
+declare     var_1CC                 u16         0x1CC
+declare     var_1CE                 u8          0x1CE
+declare     var_1CF                 u8          0x1CF
+declare     var_1D0                 u16         0x1D0
+declare     var_1D3                 u8          0x1D3
```
Comment (elmarsan) [inline address_map.txt:156]: `damage_cnt` ?

**Comment by elmarsan** [comment]
We know `em_20` is the dog. Should we replace `em_20` by `EnemyDog` or similar?

File: address_map.txt [hunk 5]
```diff
@@ -118,11 +124,54 @@ declare     tpage                   u8
...
+declare     unk_z                   s16         0x11A
+declare     var_14A                 u8          0x14A
+declare     var_14C                 u8          0x14C
+declare     var_150                 u8          0x150
+declare     var_154                 u16         0x154
+declare     var_158                 u16         0x158
+declare     var_15A                 u16         0x15A
+declare     var_16B                 u8          0x16B
+declare     kage                    Kage*       0x16C
+declare     var_17C                 s32         0x17C
```
Comment (elmarsan) [inline address_map.txt:144]: `pSeq_t_ptr` ?

**Comment by elmarsan** [comment]
> We know `em_20` is the dog. Should we replace `em_20` by `EnemyDog` or similar?

Is the same for everything we know, like licke, spider, etc

File: address_map.txt [hunk 6]
```diff
@@ -118,11 +124,54 @@ declare     tpage                   u8
 declare     clut                    u8
 declare     nFloor                  u8
 declare     parts_num               u8
-declare     pKan_t_ptr              s32
+declare     pKan_t_ptr              void*
 declare     water                   s16
 declare     type                    u8
 declare     var_10F                 u8
 
+struct      EnemyEntity
```
Comment (elmarsan) [inline address_map.txt:132]: I think we know lot's of these variables. So we should be able to rename them.

File: src/scd.cpp [hunk 7]
```diff
@@ -156,6 +159,26 @@ namespace openre::scd
...
+        uint8_t pad_01;
+        uint8_t id;
+        uint8_t type;
+        uint8_t pose;
+        uint8_t behaviour;
+        uint8_t floor;
+        uint8_t soundBank;
+        uint8_t texture;
+        uint8_t globalId;
+        int16_t x;
```
Comment (elmarsan) [inline src/scd.cpp:174]: Can x, y and z be `Vec16` ?

Comment (elmarsan) [reply src/scd.cpp:174]: It could simplify the `em->pos` and `em->old_pos` assignments.

File: src/scd.cpp [hunk 8]
```diff
@@ -522,6 +545,168 @@ namespace openre::scd
...
+        em->var_1F4 = 0;
+        if (em->id >= 0x40)
+            em->sc_id = 0x80;
+        else
+            em->sc_id = 0x04;
+        em->var_1E8 = 0;
+
+        uint16_t* atd = (uint16_t*)&em->atd;
+        atd[0x08] = 0;
+        atd[0x0A] = 64006;
```
Comment (elmarsan) [inline src/scd.cpp:643]: -1530

File: src/scd.cpp [hunk 9]
```diff
@@ -522,6 +545,168 @@ namespace openre::scd
...
+            gGameTable.c_id = em->id;
+            gGameTable.c_model_type = em->var_1CF;
+            gGameTable.c_em = em;
+            auto kind = em_kind_search(em->id);
+            if (kind != gGameTable.c_kind)
+                em->var_1CF &= ~0x80;
+            gGameTable.c_kind = kind;
+
+            auto emdResult = emd_load(em->id, em, gGameTable.mem_top);
+            if (gGameTable.ctcb->var_13 != 0)
```
Comment (elmarsan) [inline src/scd.cpp:680]: I'm not sure if this conditional is ok.

File: src/scd.cpp [hunk 10]
```diff
@@ -522,6 +545,168 @@ namespace openre::scd
...
+        auto parts = partswork_set(em, gGameTable.mem_top);
+        gGameTable.mem_top = partswork_link(em, parts, em->pKan_t_ptr, 0);
+        sa_dat_set(em, em->pSa_dat);
+        if (check_flag(FlagGroup::Status, FG_STATUS_MIRROR))
+            gGameTable.mem_top = mirror_model_cp(em, gGameTable.mem_top);
+
+        em->var_14C = 0;
+        em->var_158 = opcode->animation;
+        em->var_15A = opcode->var_14;
+        if (em->var_10F & 0x40)
```
Comment (elmarsan) [inline src/scd.cpp:704]:
```suggestion
        if ((em->var_10F & 0x40) == 0)
```

File: src/scd.cpp [hunk 11]
```diff
@@ -522,6 +545,168 @@ namespace openre::scd
...
+            em->var_1C0 = 0x92;
```
Comment (elmarsan) [inline src/scd.cpp:705]:
```suggestion
            em->var_1C0 = -110;
```

**Review by elmarsan — CHANGES_REQUESTED** [review]
Some minor changes and doubts

File: address_map.txt (hunk 6 above)
Comment (IntelOrca) [reply address_map.txt:132]: I am wondering whether PlayerEntity and EnemyEntity need to inherit a new shared base class.

Comment (elmarsan) [reply address_map.txt:132]: I think so, before introducing interop/address_map it was like that, right?

**Pull request merged**